### ISSUE TITLE
Docker module does not respect docker daemon's --ip setting to define a default binding host for published ports

### ISSUE THREAD

**Issue by SAKUJ0** [ansible/ansible-modules-core#2982]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

_docker module
##### ANSIBLE VERSION

N/A
##### SUMMARY

The title says all.

When you launch docker daemon with the --ip=10.8.0.1 setting (the default IP address of an OpenVPN server with a tun0 device), issuing a docker run with a publish setting such as `-P` or `-p 80:80` will bind the published port to the specified ip address `10.8.0.1:80->80`.

However, with the ansible module, we have to specify the bind address regardless of `--ip=address` setting of the daemon. Otherwise the result will be `0.0.0.0:80->80`.

I believe this is likely a follow-up of how https://github.com/ansible/ansible-modules-core/issues/605 has been fixed - that is just a guess, though.

**Comment by SAKUJ0** [comment]
For posterity, this is on ubuntu with everything according to documentation.

```
Containers: 6
 Running: 2
 Paused: 0
 Stopped: 4
Images: 4
Server Version: 1.10.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 92
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.0-27-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.24 GiB
```

```
Client:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 18:41:30 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 18:41:30 2016
 OS/Arch:      linux/amd6
```

```
ansible 2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = None
```

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by code-and-such** [comment]
I think I have a related problem ... Binding works properly when not giving a net (--net parameter). If a net is created the bind will be done on 0.0.0.0. Is this a different problem?

# Setup:

Created loopback @ 192.168.100.1
Starting docker deamon with --ip 192.168.100.1

## Working

docker run -d -p 1000:1000 myimage  -> works properly; bound to 192.168.100.1:1000

## Not working

docker network create mynet
docker run -d -p 1000:1000 myimage -> whoopsie; bound to 0.0.0.0:1000

# EDIT

Got it working with 
docker network create -o "com.docker.network.bridge.host_binding_ipv4=192.168.100.1" mynet

Should I remove this comment? It seems to be out of scope.

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chouseknecht** [comment]
FWIW, this was addressed [here](https://github.com/ansible/ansible-modules-core/commit/74dbb61b8d3c6cb6073a6d9c1effe6608d3e5bf3) in docker_container. Note that the docker module is being deprecated and replaced by docker_container. 

```
published_ports:
    description:
      - List of ports to publish from the container to the host.
      - "Use docker CLI syntax: C(8000), C(9000:8000), or C(0.0.0.0:9000:8000), where 8000 is a
        container port, 9000 is a host port, and 0.0.0.0 is a host interface."
      - Container ports must be exposed either in the Dockerfile or via the C(expose) option.
      - A value of ALL will publish all exposed container ports to random host ports, ignoring
        any other mappings.
      - If C(networks) parameter is provided, will inspect each network to see if there exists
        a bridge network with optional parameter com.docker.network.bridge.host_binding_ipv4.
        If such a network is found, then published ports where no host IP address is specified 
        will be bound to the host IP pointed to by com.docker.network.bridge.host_binding_ipv4.
        Note that the first bridge network with a com.docker.network.bridge.host_binding_ipv4 
        value encountered in the list of C(networks) is the one that will be used.
```

If you create a bridge network with optional parameter com.docker.network.bridge.host_binding_ipv4 and attach the container to the network via the _networks_ parameter, then ports will be exposed with the correct host IP.

I don't believe there is a way for the module to know the --ip value set when the Docker daemon started. We could remove the '0.0.0.0' as the default IP from the module, and ports would bind to the expected IP. However, the problem then becomes comparing an existing container config vs. the a requested config containing no IP address.

**Comment by ratemyduck** [comment]
Actually I found one: 

Removing the files in /var/lib/docker/network/files and restarting the deamon fixes the issue.

**Comment by chouseknecht** [comment]
wontfix

As mentioned above existing docker.py module is being deprecated. Please try out the latest version of the docker_cotainer module. If docker_container fails to perform as expected, please open an issue for docker_container.
